### ISSUE TITLE
How resilient are model2vec and potions with reduced precision?

### ISSUE THREAD

**Issue by thiswillbeyourgithub** [MinishLab/model2vec#256]
Hi,

I'm interested in super low latency to hash search queries into bins better than my current setup (using mmhash with a modulo on each token of the search query). This could even be done on the client side.

I hoped semhash could be used for that but apparently it's only for deduplication and not hashing despite the name.

So I was wondering about the use of your framework to create binary vector embeddings.

For context see https://simonwillison.net/2024/Nov/11/binary-vector-embeddings/

> Vector embeddings by themselves are pretty neat. Binary quantized vector embeddings are extra impressive. In short, they can retain 95+% retrieval accuracy with 32x compression and ~25x retrieval speedup

I did some basic tests to try SentenceTransformers's built in "precision" argument as well as basic binary masks.
(I had to cast as an array because the similarity functions usually crash with something other than float)

```python

import numpy as np
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("minishlab/potion-multilingual-128M", device="cpu")

sentences = [
    "The weather is lovely today.",
    "It's so sunny outside!",
    "He drove to the stadium."
]

embeddings = model.encode(sentences)
similarities = model.similarity(embeddings, embeddings)
print("Reference:")
print(similarities[0])
print(embeddings.dot(embeddings.T))

for prec in ["int8", "uint8", "binary", "ubinary"]:
    embeddings = np.array(model.encode(sentences, precision=prec)).astype(np.float32)
    similarities = model.similarity(embeddings, embeddings)
    print(f"{prec} precision:")
    print(similarities[0])
    print(embeddings.dot(embeddings.T))

embeddings = model.encode(sentences)
embeddings[embeddings >= 0] = 1
embeddings[embeddings < 0] = 0
similarities = model.similarity(embeddings, embeddings)
print("Naive binary:")
print(similarities[0])
print(embeddings.dot(embeddings.T))


```

Here is the python output:

```
Reference:
tensor([1.0000, 0.5073, 0.0911])
[[1.0000005  0.50730217 0.09111133]
 [0.50730217 1.0000001  0.13269816]
 [0.09111133 0.13269816 1.0000004 ]]
Computing int8 quantization buckets based on 3 embeddings. int8 quantization is more stable with `ranges` calculated from more embeddings or a `calibration_embeddings` that can be used to calculate the buckets.
int8 precision:
tensor([ 1.0000, -0.2519, -0.5258])
[[ 3225834.  -761411. -1762872.]
 [ -761411.  2831506. -1629246.]
 [-1762872. -1629246.  3484130.]]
Computing uint8 quantization buckets based on 3 embeddings. uint8 quantization is more stable with `ranges` calculated from more embeddings or a `calibration_embeddings` that can be used to calculate the buckets.
uint8 precision:
tensor([1.0000, 0.4504, 0.3125])
[[7009892. 3107044. 2322682.]
 [3107044. 6789588. 2541319.]
 [2322682. 2541319. 7883017.]]
binary precision:
tensor([ 1.0000,  0.1667, -0.1769])
[[161766.  22996. -31132.]
 [ 22996. 117690. -21985.]
 [-31132. -21985. 191353.]]
ubinary precision:
tensor([1.0000, 0.8054, 0.7379])
[[553446. 477396. 523236.]
 [477396. 634810. 595103.]
 [523236. 595103. 908409.]]
Naive binary:
tensor([1.0000, 0.5986, 0.4849])
[[113.  70.  59.]
 [ 70. 121.  67.]
 [ 59.  67. 131.]]
```

As you can see from the warning, SentenceTransformers' 'precision' arg ideally calibrates to the embeddings so I was wondering if you would be interested in running some benchmarks for this kind of usecase. Importantly, once the calibration is done, you can just pass the thresholds as argument for everyone else's benefit.

Basically I believe potion and m2v are great contenders to have a low footprint, ultra fast, compact model with compact outputs that would be a great fit for many web based use case.

**Comment by thiswillbeyourgithub** [comment]
Hi,

Thanks a lot for the answer.

I took a deeper look myself with a very crude testing code:

```python
from mteb import MTEB
from sentence_transformers import SentenceTransformer
import json
from pathlib import Path
from pprint import pprint


model_name = "minishlab/potion-multilingual-128M"
model = SentenceTransformer(model_name, device="cuda")


evaluation = MTEB(
    tasks=[
        # tasks chosen at complete random
        "PublicHealthQA",
        "BrightRetrieval",
        "Banking77Classification",
        "RestaurantReviewSentimentClassification.v2",
        "SpanishNewsClassification",
    ],
)
for prec in ["reference", "int8", "uint8", "binary", "ubinary"]:
    results = evaluation.run(
        model,
        output_folder=f"results/{prec}_{model_name}",
        encode_kwargs={"precision": prec} if prec != "reference" else {},
    )



files = [f for f in Path("results").rglob("*.json") if f.name != "model_meta.json"]

data = {}
scores = {}
times = {}
for f in files:
    loaded = json.loads(f.read_text())
    # model_data = json.loads(Path(f.parent / "model_meta.json").read_text())

    task = f.stem
    if task not in data:
        data[task] = {}
        scores[task] = {}
        times[task] = {}

    run_name = f.parent.parent.parent.parent.name
    assert run_name not in data[task], run_name
    data[task][run_name] = loaded

    times[task][run_name] = loaded["evaluation_time"]

    if "test" in loaded["scores"]:
        scores[task][run_name] = loaded["scores"]["test"][0]["main_score"]
    elif "train" in loaded["scores"]:
        scores[task][run_name] = loaded["scores"]["train"][0]["main_score"]
    elif "standard" in loaded["scores"]:
        scores[task][run_name] = loaded["scores"]["standard"][0]["main_score"]
    else:
        breakpoint()
    # if "scores_per_experiment" in scores[task][run_name]:
    #     del scores[task][run_name]["scores_per_experiment"]

pprint(scores)
pprint(times)
breakpoint()
```


Here are the result:

For the scores:

```
{'Banking77Classification': {'binary_minishlab': 0.286104,
                             'int8_minishlab': 0.723474,
                             'reference_minishlab': 0.685292,
                             'ubinary_minishlab': 0.307792,
                             'uint8_minishlab': 0.731136},
 'BrightRetrieval': {'binary_minishlab': 0.00445,
                     'int8_minishlab': 0.04665,
                     'reference_minishlab': 0.09851,
                     'ubinary_minishlab': 0.0046,
                     'uint8_minishlab': 0.04977},
 'PublicHealthQA': {'binary_minishlab': 0.21349,
                    'int8_minishlab': 0.55758,
                    'reference_minishlab': 0.47392,
                    'ubinary_minishlab': 0.1557,
                    'uint8_minishlab': 0.51265},
 'RestaurantReviewSentimentClassification.v2': {'binary_minishlab': 0.502002,
                                                'int8_minishlab': 0.56084,
                                                'reference_minishlab': 0.585107,
                                                'ubinary_minishlab': 0.506494,
                                                'uint8_minishlab': 0.591992},
 'SpanishNewsClassification': {'binary_minishlab': 0.509912,
                               'int8_minishlab': 0.867383,
                               'reference_minishlab': 0.873779,
                               'ubinary_minishlab': 0.519824,
                               'uint8_minishlab': 0.863086}}
```

For the evaluation duration:
```
{'Banking77Classification': {'binary_minishlab': 4.333920240402222,
                             'int8_minishlab': 5.485781908035278,
                             'reference_minishlab': 6.0497870445251465,
                             'ubinary_minishlab': 4.629784107208252,
                             'uint8_minishlab': 7.501502275466919},
 'BrightRetrieval': {'binary_minishlab': 1281.3917224407196,
                     'int8_minishlab': 1294.1054031848907,
                     'reference_minishlab': 1261.5905675888062,
                     'ubinary_minishlab': 1307.40047621727,
                     'uint8_minishlab': 1297.522028684616},
 'PublicHealthQA': {'binary_minishlab': 1.6070215702056885,
                    'int8_minishlab': 1.6569535732269287,
                    'reference_minishlab': 1.661961317062378,
                    'ubinary_minishlab': 1.7295451164245605,
                    'uint8_minishlab': 1.6639628410339355},
 'RestaurantReviewSentimentClassification.v2': {'binary_minishlab': 9.817237615585327,
                                                'int8_minishlab': 10.097695350646973,
                                                'reference_minishlab': 10.577326774597168,
                                                'ubinary_minishlab': 10.242697477340698,
                                                'uint8_minishlab': 10.280569791793823},
 'SpanishNewsClassification': {'binary_minishlab': 14.943988800048828,
                               'int8_minishlab': 16.351204872131348,
                               'reference_minishlab': 17.424144506454468,
                               'ubinary_minishlab': 15.7865309715271,
                               'uint8_minishlab': 15.288179636001587}}
```


So basically:
- the duration are so similar that I'm thinking there might be something wrong with how timing is measured
- binary and ubinary appear rubbish on the surface, but if they could be optimized for speed they would still be a  huge improvement over non semantic hashing + modulo
- the int8 and uint8 are similar (sometimes better?!) than the default, and usually quicker
-

My conclusion is that I am not convinced that there is no use for a very optimized cpu only binary embedder. It seems like a useful technological block, and one where there is maybe little competition.


Another question came to my mind today: given that your model2vec is done by distillation IIRC, doesn't that paves the way for quantization aware distillation? I'd be surprised if there was not a substantial amount of accuracy to regain when going binary with this kind of setup.

Interested in your thoughts on this.

**Comment by stephantul** [comment]
Hello!

Thanks for doing the benchmarks. Binary quantization will never lead to a speed-up, and, in fact, is probably a tiny bit slower than just creating means using Model2Vec. What binary quantization speeds up is optimized retrieval speed and storage, both of which you don't measure here. Your benchmark results mirror mine.

It's interesting that int8/uint8 sometimes is better, but I expect them to average out over the whole MMTEB test suite, and would be surprised if either is much better on average across all models.

Regarding your comments about hashing: if you have a vector embedding and want a real semantic hash, the best way to go about this is to use a product quantizer. This is basically what many vector databases do under the hood anyway, so in a sense the hashing part of what you desire is accomplished implicitly by simply inserting your vectors into a vector database. If you want transparent semantic hashes with a predefined number of bits, you could look into something like nanopq: https://github.com/matsui528/nanopq

**Comment by thiswillbeyourgithub** [comment]
I thank you very much for taking the time to teach me all this.

I was about to close this issue but actually there's still an unanswered question :)

> Another question came to my mind today: given that your model2vec is done by distillation IIRC, doesn't that paves the way for quantization aware distillation? I'd be surprised if there was not a substantial amount of accuracy to regain when going binary with this kind of setup.

**Comment by stephantul** [comment]
@thiswillbeyourgithub I don't know, one thing that could be done I guess is to quantize the output of the given sentence transformer while distilling with reference to the corpus of items you'd like to featurize. But that's something we already do when performing int8 quantization. We also perform PCA over the reduced space, which you can see as a form of shaping of the space into one that is more useful given the embeddings you are creating.

**Comment by thiswillbeyourgithub** [comment]
Reading a bit of the code I think I meant more something like this:
1. You create the StaticModel by using the embeddings from the vocabulary of the model.
2. Then you sort of create again embeddings form the whole vocabulary but this time the embeddings are from the StaticModel, not the original model.

3. You create a BinaryStaticModel (BSM for short) by setting to 1 everything above median and 0 everything below the median.

4. You take a look at the value in the StaticModel that are within 10% of the median value. All other values from the BSM will be frozen.

5. You then create all the embeddings from the vocabulary using that BSM model.

6. I don't know the exact loss to use to try to align the BSM with the SM but maybe you binarize the output from SM, and XOR with the output from BSM then sum and normalize to number of dimension?

7. You then backpropagate only on the value that are within 10% of the median.

8. Rinse and repeat a customizable number of times, and with a customizable % of deviation from the median.

If that works you'd have a BinaryStaticModel that is even smaller, outputs something super fast to compare (just XOR), and has results closer to the StaticModel, which is already not very far from the original model.

Or maybe I am totally delirious. In any case I would appreciate any form of criticism on this, but would understand if you decide not to even engage as I'm completely out of my domain and don't want to waste your time.

And in any case I'll read your RSS.

Thanks and bye.

**Comment by stephantul** [comment]
I didn't mean to imply that you were wrong or delirious, but I don't necessarily see a use case for a binarized model. Binarized output embeddings are easy to create with a reference corpus, as you indicated above, and work well with retrieval over large corpora. Because retrieval over a very large corpus is (one of the?) only use case(s) for a binarized model, I think these go hand in hand.

Also note that having binarized weights doesn't mean you will have binarized output embeddings. You will still need a "binarization" stage to turn the binarized weights into binarized output embeddings, just as before. By this I mean to say that you are probably still better off having real-valued weights and then binarizing these into output embeddings, instead of having binary weights.

Nevertheless, if you are interested in trying, you can easily binarize an embedding model using backpropagation by training a model with a really high-temperature sigmoid. You can take static embedding model, create a high-dimensional matrix with the same number of rows, but many many more columns, and then simply perform knowledge distillation. The extremely high temperature sigmoid still allows gradients to backpropagate (using a step function like > median doesn't work). After training, you simply forward propagate everything through the high temp sigmod again, and round.

**Comment by thiswillbeyourgithub** [comment]
Thanks again.

(Instead of very large corpus I was more thinking about embedded so with low ram and low disks.)
